src/diagnostics.rs

```diff
@@ -0,0 +1,633 @@
+    type__incompatible_directaccess_variable,
+    type__incompatible_directaccess_range,
+    type__expected_literal,
+
+    //codegen related
+    codegen__general,
+    codegen__missing_function,
+}
+
+impl Diagnostic {
+    pub fn syntax_error(message: &str, range: SourceRange) -> Diagnostic {
+        Diagnostic::SyntaxError {
+            message: message.to_string(),
+            range,
+            err_no: ErrNo::syntax__generic_error,
+        }
+    }
+
+    pub fn unexpected_token_found(expected: &str, found: &str, range: SourceRange) -> Diagnostic {
+        Diagnostic::SyntaxError {
+            message: format!(
+                "Unexpected token: expected {} but found {}",
+                expected, found
+            ),
+            range,
+            err_no: ErrNo::syntax__unexpected_token,
+        }
+    }
+
+    pub fn unexpected_initializer_on_function_return(range: SourceRange) -> Diagnostic {
+        Diagnostic::SyntaxError {
+            message: "Return types cannot have a default value".into(),
+            range,
+            err_no: ErrNo::syntax__unexpected_token,
+        }
+    }
+
+    pub fn return_type_not_supported(pou_type: &PouType, range: SourceRange) -> Diagnostic {
+        Diagnostic::SyntaxError {
+            message: format!(
+                "POU Type {:?} does not support a return type. Did you mean Function?",
+                pou_type
+            ),
+            range,
+            err_no: ErrNo::pou__unexpected_return_type,
+        }
+    }
+
+    pub fn function_unsupported_return_type(data_type: &DataTypeDeclaration) -> Diagnostic {
+        Diagnostic::SyntaxError {
+            message: format!(
+                "Data Type {:?} not supported as a function return type!",
+                data_type
+            ),
+            range: data_type.get_location(),
+            err_no: ErrNo::pou__unsupported_return_type,
+        }
+    }
+
+    pub fn function_return_missing(range: SourceRange) -> Diagnostic {
+        Diagnostic::SyntaxError {
+            message: "Function Return type missing".into(),
+            range,
+            err_no: ErrNo::pou__missing_return_type,
+        }
+    }
+
+    pub fn missing_function(location: SourceRange) -> Diagnostic {
+        Diagnostic::SyntaxError {
+            message: "Cannot generate code outside of function context.".into(),
+            range: location,
+            err_no: ErrNo::codegen__missing_function,
+        }
+    }
+
+    pub fn missing_token(epxected_token: &str, range: SourceRange) -> Diagnostic {
+        Diagnostic::SyntaxError {
+            message: format!("Missing expected Token {}", epxected_token),
+            range,
+            err_no: ErrNo::syntax__missing_token,
+        }
+    }
+
+    pub fn missing_action_container(range: SourceRange) -> Diagnostic {
+        Diagnostic::SyntaxError {
+            message: "Missing Actions Container Name".to_string(),
+            range,
+            err_no: ErrNo::undefined,
```

Comment:
I would go with a defined error here since we can now derive an actions container's name based on its order in a file.

Reply:
I also don't know if this was just the rename of the file, if it was we could handle it later